master/internal/resourcemanagers/fitting_test.go

```diff
@@ -14,232 +14,337 @@ func TestIsViable(t *testing.T) {
 	assert.Assert(t, !isViable(req,
-		newMockAgent(t, system, "agent4", 1, ""), slotsSatisfied))
+		newFakeAgentState(t, system, "agent2", "", 1, 0, nil, 0), slotsSatisfied))
+	assert.Assert(t, !isViable(req,
+		newFakeAgentState(t, system, "agent4", "", 1, 0, nil, 0), slotsSatisfied))
 }
 
-func TestFindFit(t *testing.T) {
-	system := actor.NewSystem(t.Name())
-
+func TestFindFits(t *testing.T) {
 	type testCase struct {
-		Name                string
-		SlotsNeeded         int
-		AgentCapacities     []int
-		AgentOccupiedSlots  []int
-		AgentLabels         [2]string
-		FittingMethod       SoftConstraint
-		ExpectedAgentFit    int
-		TaskLabel           string
-		FittingRequirements FittingRequirements
+		Name          string
+		Task          AllocateRequest
+		Agents        []*mockAgent
+		FittingMethod SoftConstraint
+
+		ExpectedAgentFit int
 	}
 
 	testCases := []testCase{
 		{
-			Name:             "2-slot multiple fits",
-			SlotsNeeded:      2,
-			AgentCapacities:  []int{2, 4},
+			Name: "0-slot multiple fits, idle agents",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent1", "", 4, 0, nil, 0),
+				newMockAgent("agent2", "", 4, 0, nil, 0),
+			},
 			FittingMethod:    BestFit,
 			ExpectedAgentFit: 0,
-			FittingRequirements: FittingRequirements{
-				SingleAgent: false,
+		},
+		{
+			Name: "0-slot multiple fits, idle agents, out-of-order",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent2", "", 4, 0, nil, 0),
+				newMockAgent("agent1", "", 4, 0, nil, 0),
 			},
+			FittingMethod:    BestFit,
+			ExpectedAgentFit: 1,
 		},
 		{
-			Name:             "1-slot multiple fits",
-			SlotsNeeded:      1,
-			AgentCapacities:  []int{1, 4},
+			Name: "0-slot multiple fits, in-use agents",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent1", "", 4, 0, nil, 2),
+				newMockAgent("agent2", "", 4, 0, nil, 2),
+			},
 			FittingMethod:    BestFit,
 			ExpectedAgentFit: 0,
 		},
 		{
-			Name:             "1-slot out-of-order multiple fits",
-			SlotsNeeded:      1,
-			AgentCapacities:  []int{4, 1},
+			Name: "0-slot multiple fits, in-use agents, out-of-order",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent2", "", 4, 0, nil, 2),
+				newMockAgent("agent1", "", 4, 0, nil, 2),
+			},
 			FittingMethod:    BestFit,
 			ExpectedAgentFit: 1,
 		},
 		{
-			Name:             "4-slot single fit",
-			SlotsNeeded:      4,
-			AgentCapacities:  []int{4, 1},
+			Name: "0-slot multiple fits, max zero slot containers",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent1", "", 4, 0, newIntPtr(2), 0),
+				newMockAgent("agent2", "", 4, 0, newIntPtr(4), 2),
+			},
 			FittingMethod:    BestFit,
 			ExpectedAgentFit: 0,
```

Comment:
Why do we expect this to get scheduled on `agent1`?

Reply:
Because they both have limits and the fitting score is calculated by 1/(1+maxCpuContainersPerAgent-agent.numZeroSlotContainers). `Agent1` and `Agent2` got the same score but `Agent1` has a smaller hash distance.